### PR TITLE
Audit Key Deduper

### PR DESCRIPTION
**Pull request by iBotPeaches**
We are finding in a project that the audit log has grown to near 11gb, because each key logged is like

 * `data.1.name, data.2.name, data.3.name`

The key is implicit so does not tell us anything and we already know which models were interacted with, so the purpose of this is solely to know which fields of data.

So we can dedupe/sort and clean this up so it becomes:

 * `name`
 
We basically take the key after the last `.`, pluck out the duplicates, sort them and hash. This will keep a clean database of unique combinations of possible key types.

---

Next steps is to introduce a migration to cleanup existing data.

### CODE REVIEW

**Comment by iBotPeaches** [comment]
Putting this on draft. Discovered relations break this as `data.1.related.bar` would become just `bar`.

**Review by bradtoad — APPROVED** [review]

**Review by PhilSanders — APPROVED** [review]

File: database/migrations/2023_01_13_000005_correct_inefficient_auditkeys.php [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+use Illuminate\Database\Migrations\Migration;
+use Illuminate\Support\Facades\DB;
+use Sourcetoad\Logger\Helpers\DataArrayParser;
+use Sourcetoad\Logger\Models\AuditKey;
+
+return new class extends Migration
+{
+    public function up(): void
+    {
+        AuditKey::query()->chunkById(1000, function (Collection $auditKeys) {
```
Comment (justinvoelkel) [inline database/migrations/2023_01_13_000005_correct_inefficient_auditkeys.php:13]: Any chance we run into large data sets better served by using `cursor`?

**Review by justinvoelkel — APPROVED** [review]

File: database/migrations/2023_01_13_000005_correct_inefficient_auditkeys.php (hunk 1 above)
Comment (iBotPeaches) [reply database/migrations/2023_01_13_000005_correct_inefficient_auditkeys.php:13]: 100k is most and I peeked this thread: https://stackoverflow.com/a/53736557 and thought 1000 chunking looked the best.

Comment (justinvoelkel) [reply database/migrations/2023_01_13_000005_correct_inefficient_auditkeys.php:13]: Eh I guess thats the point of chunking - nvm.

**Pull request merged**